### ISSUE TITLE
Nip-17 "Chat Room" definition is too restrictive : clients and users need more flexibility.

### ISSUE THREAD

**Issue by manimejia** [nostr-protocol/nips#1297]
The "Chat Room" section is unnecessarily specific in what "defines a chat room". Clients and users SHOULD be given more flexibility to VIEW and interact with message threads as they see fit. 

https://github.com/nostr-protocol/nips/blob/604ab3cee1d6ad70c797912de33f487c82d9fc47/17.md?plain=1#L38-L44

As written, this specification currently REQUIRES a client to display an empty "room" EVERY TIME a new user is added or removed from a message 'reply'. Given the permissionless nature of "being added" to a "thread" (any user can ad or remove another at any time) such a rigid "chat room definition" will likely result  in fractured and hard to follow conversations. Users and clients should have the flexibility to adapt.

Here's an idea of how this could be rewritten :

---

 ## Chat Rooms 
  
A "Chat Room" is a rendering of private messages as a single thread. Clients SHOULD render private messages this way, and  SHOULD offer users flexibility in how (each) room is defined. A "Chat Room"  MAY be defined by any number of tags found in the`Kind 14` messages retrieved for a given user.  Clients MAY display a new room (with a clean message history) when any ONE or MORE of these tags in a single message are DIFFERENT from other messages. 

- `p` tags : Required `p` tags define the participants "viewers" of a single message. Any "member" of a current message MAY "add" or "remove" participants from a message reply. 
- `e` tags : An optional `e` tag MAY refer to another (decrypted `kind:14`) event `id` as an ancestor (that this message is a 'reply' to). A user MAY receive a reply to an ancestor message that IS NOT available, depending on 'Chat Room' settings.
- `subject` tags : An optional `subject` tag defines the current name/topic of the conversation for any given message. Any member can change the subject for a message sent. Clients SHOULD NOT add a `subject` tag if the user has not changed it. 

Some EXAMPLES of possible rules which MAY define a new "Chat Room" :

- message has a single differing (added or removed) a `p` tag
- message has a different `subject` tag
- message is NOT a reply (`e` tag) to a message in the room.
- user XXX has changed `subject` tag
- user XXX has added or removed a `p` tag
- user XXX who set `subject` tag is not included in message `p` tags

Clients and users SHOULD have the freedom to adapt "Chat Room" definitions to suit their individual needs.

**Comment by alexgleason** [comment]
See also: #1229

**Comment by vitorpamplona** [comment]
The main issue is how to make sure users understand which user received which message. When you add a new user to a chat, it's assumed that the user can see all past conversations, which is not true for NIP-17. Similarly, if a user replies with a sub-group of the group and the message is shown in the same thread, there is no clear visual for which users can see which messages create a bunch of confused users, especially in 5+ groups. The group itself becomes a bunch of branched conversions with separate subgroups not being able to see each other even though their messages appear in the same conversation flow.

To simplify this mess, we made everyone change rooms when new users are added or removed. In that way, it becomes clear which conversations are visible to which users at all times. 

This is very similar to how Slack does group DMs.

**Comment by vitorpamplona** [comment]
> These decisions should NOT be part of NIP specifications. UX matters like this SHOULD be ironed out by implementers, with the NIP specifying ONLY guidelines for such experiments.

Most NIPs offer a specific type of implementation with semantical decisions built-in. It helps any new implementer to become interoperable without having to figure out how clients do.... Chat rooms... in this case. 

However, we are happy to change the NIP if an implementation can solve this without confusing everyone.

**Comment by vitorpamplona** [comment]
What happens when other Clients start replying to those chat rooms, adding and removing people, and changing the subject at random? 

Keep in mind that for larger group chats, say the advocate wants to add all of his invites into a single group, @staab has created #875, which does include the idea of an admin and moderation to add and remove people from groups. In those cases, the chat group is whatever the advocate wants it to be and it doesn't allow the invitees to interfere with the group, like NIP-17 does.

**Comment by vitorpamplona** [comment]
> Indeed, interoperability may be improved by allowing clients (and users) the flexibility to "be creative with this novel architecture" to "see what works".

Flexibility generally doesn't improve interoperability. The direction generally goes in the opposite way: the more the specific, the easier it becomes to be interoperable. 

Now, NIPs don't block "be creative with this novel architecture" to "see what works". You can still do what you want, just keep in mind that other clients will remap your chatrooms in the way they see fit. So, if things are shown as a chatroom in your client but as separate rooms in other clients, users might be very confused. 

The NIP is just a guide to what you should expect to see out there. If you can create your client in your way still make it so users are not confused when they switch, go for it. We can adapt the NIP when we see how the client actually works.

**Comment by vitorpamplona** [comment]
NIPs don't dictate anything. They just reflect the implemented reality out there. They change as new clients come in. 

So, never feel like you can't do something just because a NIP is saying so. Develop your client and the NIP will have to change to make sure other devs know who things are being used.
